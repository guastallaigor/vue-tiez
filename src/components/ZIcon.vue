<template>
  <component :is="name" :size="size" :dark="dark"/>
</template>

<script>
export default {
  name: 'z-icon',
  components: {
    add: () => import('./icons/add'),
    arrow: () => import('./icons/arrow'),
    'arrow-double': () => import('./icons/arrow-double')
  },
  props: {
    name: {
      type: String,
      required: true
    },
    size: {
      type: String,
      default: '20px'
    },
    dark: {
      type: Boolean,
      default: false
    }
  }
}
</script>

<style lang="scss">
svg.z-icon {
  shape-rendering: crispEdges;

  circle {
    transition: all 0.2s ease 0s;
  }

  &:hover circle {
    stroke: rgb(153, 153, 153);
  }
}
</style>
