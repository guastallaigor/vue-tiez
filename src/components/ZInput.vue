<template>
  <div class="z-input" :style="label ? 'display:inline-flex;width:initial' : ''">
    <span class="label" :class="{'dark': dark}" v-if="label">{{ label }}</span>
    <div class="wrapper" :class="getClasses">
      <span class="icon" v-if="icon"><z-icon :name="icon" :dark="dark" /></span>
      <div>
        <input
          ref="input"
          autocapitalize="off"
          autocomplete="off"
          autocorrect="off"
          :placeholder="placeholder"
          :type="type"
          :disabled="disabled"
          :maxlength="maxLength"
          v-bind="$attrs"
          @blur="onBlur"
          @focus="onFocus"
          v-model="valueComp"
          >
      </div>
    </div>
  </div>
</template>

<script>
export default {
  name: 'z-input',
  components: {
    ZIcon: () => import('./ZIcon')
  },
  props: {
    autoFocus: {
      type: Boolean,
      default: false
    },
    dark: {
      type: Boolean,
      default: false
    },
    disabled: {
      type: Boolean,
      default: false
    },
    error: {
      type: Boolean,
      default: false
    },
    icon: {
      type: String,
      default: ''
    },
    label: {
      type: String,
      default: ''
    },
    placeholder: {
      type: String,
      default: ''
    },
    type: {
      type: String,
      default: 'text'
    },
    value: {
      type: [String, Number],
      default: ''
    },
    maxLength: {
      type: [String, Number],
      default: ''
    }
  },
  data: () => ({
    focused: false
  }),
  computed: {
    getClasses () {
      const { disabled, dark, focused, error, icon, label } = this

      return {
        disabled,
        focused,
        'inner-input': label,
        error,
        icon,
        dark
      }
    },
    valueComp: {
      get () {
        return this.value
      },
      set (val) {
        if (this.disabled) return

        if (val && val.length > this.maxLength) return

        this.$emit('input', val)
      }
    }
  },
  mounted () {
    if (this.autoFocus) {
      this.$refs.input.focus()
    }
  },
  methods: {
    onFocus (event) {
      this.$emit('focus', event)
      this.focused = true
    },
    onBlur (event) {
      this.$emit('blur', event)
      this.focused = false
    }
  }
}
</script>

<style lang="scss">
.z-input {
  .label {
    background-color: #fafafa;
    border-bottom-left-radius: 5px;
    border-bottom: 1px solid #e1e1e1;
    border-left: 1px solid #e1e1e1;
    border-top-left-radius: 5px;
    border-top: 1px solid #e1e1e1;
    color: #9b9b9b;
    font-size: 14px;
    line-height: 1;
    margin: 0;
    padding: 10px;
    pointer-events: none;
    width: initial;

    &.dark {
      background-color: #111;
      border-bottom: 1px solid #333;
      border-left: 1px solid #333;
      border-top: 1px solid #333;
    }
  }

  .wrapper {
    align-items: center;
    border-radius: 5px;
    border: 1px solid #e1e1e1;
    display: inline-flex;
    height: 37px;
    position: relative;
    transition: border 0.2s ease, color 0.2s ease;
    vertical-align: middle;
    width: 100%;
    background: #fff;

    input {
      background-color: transparent;
      border-radius: 0;
      border: none;
      box-shadow: none;
      position: relative;
      box-sizing: border-box;
      display: block;
      font-size: 14px;
      line-height: 27px;
      outline: 0;
      width: 100%;
      color: rgb(0, 0, 0);
      caret-color: rgb(0, 0, 0);
      height: 100%;
      opacity: 1;

      &:disabled {
        color: #999;
        border-radius: 5px;
        cursor: not-allowed;

        &::placeholder {
          color: #999;
        }
      }
    }

    &.inner-input {
      border-bottom-left-radius: 0;
      border-top-left-radius: 0;
      flex: 1;
    }

    &:not(.dark) {
      input {
        background: #fafafa;
      }
    }

    &.dark {
      border: 1px solid #333;
      background: #000;

      input {
        color: #fff;
        caret-color: #fff;
      }
    }

    &.disabled:not(.dark) {
      background: #fafafa;
    }

    &.focused {
      border: 1px solid #888;

      &.dark {
        border: 1px solid #999;
      }
    }

    &.error {
      border: 1px solid red;

      input {
        color: red;

        &::placeholder {
          color: red;
        }
      }

      &.focused {
        border: 1px solid red;
        color: red;
      }
    }

    .icon {
      align-items: center;
      display: flex;
      height: 100%;
      vertical-align: middle;
      padding: 0 14px;
    }

    div:first-child {
      display: block;
      margin: 4px 10px;
      position: relative;
      width: 100%;
    }
  }
}
</style>
