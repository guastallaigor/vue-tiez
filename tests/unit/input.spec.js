import { shallowMount, mount } from '@vue/test-utils'
import ZInput from '../../src/components/ZInput'
import ZIcon from '../../src/components/ZIcon'

describe('When I create the ZInput component', () => {
  const createComponent = (propsData = {}, slot) => {
    return shallowMount(ZInput, {
      propsData: propsData,
      stubs: { ZIcon },
      slots: {
        default: slot || 'ZInput'
      }
    })
  }

  it('should be a Vue instance', () => {
    const wrapper = createComponent({ value: 'test' })
    expect(wrapper.find(ZInput).isVueInstance()).toBe(true)
  })

  it('should mount the ZIcon component', (done) => {
    const wrapper = mount(ZInput, {
      propsData: { value: 'test', icon: 'add' }
    })
    const input = wrapper.find('.z-input')
    expect(input.exists()).toBe(true)
    expect(input.classes().length).toBe(1)
    expect(input.classes()).toEqual(['z-input'])
    setTimeout(() => {
      expect(wrapper.find(ZIcon).exists()).toBe(true)
      const circle = wrapper.find('.z-input > .wrapper > .icon > svg > circle')
      expect(circle.attributes().stroke).toBe('#EAEAEA')
      const line = wrapper.find('.z-input > .wrapper > .icon > svg > line')
      expect(line.attributes().stroke).toBe('#999999')
      const wrapperDiv = wrapper.find('.z-input > .wrapper')
      expect(wrapperDiv.exists()).toBe(true)
      expect(wrapperDiv.classes().length).toBe(2)
      expect(wrapperDiv.classes()).toEqual(['wrapper', 'icon'])
      done()
    })
  })

  it('should mount the ZIcon component with a dark class', (done) => {
    const wrapper = mount(ZInput, {
      propsData: { value: 'test', icon: 'add', dark: true }
    })
    const input = wrapper.find('.z-input')
    expect(input.exists()).toBe(true)
    expect(input.classes().length).toBe(1)
    expect(input.classes()).toEqual(['z-input'])
    setTimeout(() => {
      expect(wrapper.find(ZIcon).exists()).toBe(true)
      const circle = wrapper.find('.z-input > .wrapper > .icon > svg > circle')
      expect(circle.attributes().stroke).toBe('#444')
      const line = wrapper.find('.z-input > .wrapper > .icon > svg > line')
      expect(line.attributes().stroke).toBe('#999999')
      const wrapperDiv = wrapper.find('.z-input > .wrapper')
      expect(wrapperDiv.exists()).toBe(true)
      expect(wrapperDiv.classes().length).toBe(3)
      expect(wrapperDiv.classes()).toEqual(['wrapper', 'icon', 'dark'])
      done()
    })
  })

  it('should have a dark class', () => {
    const wrapper = createComponent({ value: '', dark: true })
    const inputLabel = wrapper.find('.z-input > .wrapper')
    expect(inputLabel.exists()).toBe(true)
    expect(inputLabel.classes().length).toBe(2)
    expect(inputLabel.classes()).toEqual(['wrapper', 'dark'])
  })

  it('should have a dark class and have a label', () => {
    const wrapper = createComponent({ value: '', dark: true, label: 'Label' })
    const label = wrapper.find('.z-input > .label')
    expect(label.exists()).toBe(true)
    expect(label.classes().length).toBe(2)
    expect(label.classes()).toEqual(['label', 'dark'])
  })

  it('should have a label', () => {
    const wrapper = createComponent({ value: '', label: 'Label' })
    const label = wrapper.find('.z-input > .label')
    expect(label.exists()).toBe(true)
    expect(label.classes().length).toBe(1)
    expect(label.classes()).toEqual(['label'])
    const input = wrapper.find('.z-input')
    expect(input.exists()).toBe(true)
    expect(input.classes().length).toBe(1)
    expect(input.attributes().style).toBe('display: inline-flex;')
    const wrapperDiv = wrapper.find('.z-input > .wrapper')
    expect(wrapperDiv.exists()).toBe(true)
    expect(wrapperDiv.classes().length).toBe(2)
    expect(wrapperDiv.classes()).toEqual(['wrapper', 'inner-input'])
  })

  it('should have a type text', async () => {
    const wrapper = createComponent({ value: '', type: 'text' })
    const input = wrapper.find('.z-input > .wrapper > div > input')
    expect(input.exists()).toBe(true)
    expect(input.attributes().type).toBe('text')
    input.element.value = 'test'
    input.trigger('input')
    await wrapper.vm.$nextTick()
    expect(wrapper.vm.$refs.input.value).toBe('test')
    expect(wrapper.emitted().input).toBeTruthy()
    expect(wrapper.emitted().input[0]).toEqual(['test'])
  })

  it('should have a type number', async () => {
    const wrapper = createComponent({ value: '', type: 'number' })
    const input = wrapper.find('.z-input > .wrapper > div > input')
    expect(input.exists()).toBe(true)
    expect(input.attributes().type).toBe('number')
    input.element.value = 123
    input.trigger('input')
    await wrapper.vm.$nextTick()
    expect(wrapper.vm.$refs.input.value).toBe('123')
    expect(wrapper.emitted().input).toBeTruthy()
    expect(wrapper.emitted().input[0]).toEqual(['123'])
  })

  it('should have a maxLength of 3 characters', () => {
    const wrapper = createComponent({ value: '', maxLength: 3 })
    const input = wrapper.find('.z-input > .wrapper > div > input')
    expect(input.exists()).toBe(true)
    expect(input.attributes().type).toBe('text')
    expect(input.attributes().maxlength).toBe('3')
  })

  it('should have a disable attribute and class', async () => {
    const wrapper = createComponent({ value: '', disabled: true })
    const input = wrapper.find('.z-input > .wrapper > div > input')
    expect(input.exists()).toBe(true)
    expect(input.attributes().type).toBe('text')
    expect(input.attributes().disabled).toBe('disabled')
    const wrapperDiv = wrapper.find('.z-input > .wrapper')
    expect(wrapperDiv.exists()).toBe(true)
    expect(wrapperDiv.classes().length).toBe(2)
    expect(wrapperDiv.classes()).toEqual(['wrapper', 'disabled'])
    input.element.value = 'test'
    input.trigger('input')
    await wrapper.vm.$nextTick()
    wrapper.vm.valueComp = 'other'
    expect(wrapper.vm.valueComp).toBe('')
    expect(wrapper.emitted().input).toBeFalsy()
  })

  it('should have a placeholder', () => {
    const wrapper = createComponent({ value: '', placeholder: 'Placeholder' })
    const input = wrapper.find('.z-input > .wrapper > div > input')
    expect(input.exists()).toBe(true)
    expect(input.attributes().placeholder).toBe('Placeholder')
  })

  it('should have an error class', () => {
    const wrapper = createComponent({ value: '', error: true })
    const wrapperDiv = wrapper.find('.z-input > .wrapper')
    expect(wrapperDiv.exists()).toBe(true)
    expect(wrapperDiv.classes().length).toBe(2)
    expect(wrapperDiv.classes()).toEqual(['wrapper', 'error'])
  })

  it('should emit a blur event', async () => {
    const wrapper = createComponent({ value: '' })
    const input = wrapper.find('.z-input > .wrapper > div > input')
    expect(input.exists()).toBe(true)
    expect(input.attributes().type).toBe('text')
    input.element.value = 'test'
    input.trigger('blur')
    await wrapper.vm.$nextTick()
    expect(wrapper.emitted().blur).toBeTruthy()
    expect(wrapper.vm.focused).toBe(false)
  })

  it('should emit a focus event', async () => {
    const wrapper = createComponent({ value: '', autoFocus: true })
    const input = wrapper.find('.z-input > .wrapper > div > input')
    expect(input.exists()).toBe(true)
    expect(input.attributes().type).toBe('text')
    input.trigger('focus')
    await wrapper.vm.$nextTick()
    expect(wrapper.emitted().focus).toBeTruthy()
    expect(wrapper.vm.focused).toBe(true)
  })

  it('should match snapshot', () => {
    const wrapper = createComponent({
      autoFocus: false,
      dark: false,
      disabled: false,
      error: false,
      icon: '',
      label: '',
      placeholder: '',
      type: 'text',
      value: '',
      maxLength: ''
    })

    expect(wrapper.html()).toMatchSnapshot()
  })
})
