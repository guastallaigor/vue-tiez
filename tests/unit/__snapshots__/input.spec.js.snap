// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`When I create the ZInput component should match snapshot 1`] = `
<div class="z-input">
  <!---->
  <div class="wrapper">
    <!---->
    <div><input autocapitalize="off" autocomplete="off" autocorrect="off" placeholder="" maxlength="" type="text"></div>
    <!---->
  </div>
</div>
`;
